File: src/main/java/com/zerozero/core/util/AWSS3Service.java
package com.zerozero.core.util;

import com.amazonaws.HttpMethod;
import com.amazonaws.SdkClientException;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.Headers;
import com.amazonaws.services.s3.model.CannedAccessControlList;
import com.amazonaws.services.s3.model.GeneratePresignedUrlRequest;
import com.amazonaws.services.s3.model.ObjectMetadata;
import com.amazonaws.services.s3.model.PutObjectRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.server.ResponseStatusException;

import java.io.IOException;
import java.io.InputStream;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class AWSS3Service {

    @Value("${cloud.aws.s3.bucket}")
    private String bucket;

    private final AmazonS3 amazonS3;

    public String getPreSignedUrl(String prefix, String fileName) {
        if (prefix == null || prefix.isEmpty()) {
            return null;
        }
        String filePath = createFilePath(prefix, fileName);
        GeneratePresignedUrlRequest generatePresignedUrlRequest = createGeneratePresignedUrlRequest(bucket, filePath);
        try {
            URL url = amazonS3.generatePresignedUrl(generatePresignedUrlRequest);
            return url.toString();
        } catch (SdkClientException e) {
            return null;
        }
    }

    public String getObjectUrlFromPreSignedUrl(String preSignedUrl) {
        if (preSignedUrl == null || preSignedUrl.isEmpty()) {
            return null;
        }
        try {
            URL url = new URL(preSignedUrl);
            return String.format("%s://%s%s", url.getProtocol(), url.getHost(), url.getPath());
        } catch (MalformedURLException e) {
            return null;
        }
    }

    public String uploadImage(String prefix, MultipartFile multipartFile) throws IOException {
        String fileName = createFilePath(prefix, multipartFile.getOriginalFilename());

        ObjectMetadata objectMetadata = new ObjectMetadata();
        objectMetadata.setContentLength(multipartFile.getSize());
        objectMetadata.setContentType(multipartFile.getContentType());

        amazonS3.putObject(bucket, fileName, multipartFile.getInputStream(), objectMetadata);
        return getUrl(bucket, fileName);
    }

    public List<String> uploadImages(String prefix, List<MultipartFile> multipartFiles) {
        List<String> imageUrls = new ArrayList<>();

        multipartFiles.forEach(file -> {
            String fileName = createFilePath(prefix, file.getOriginalFilename());

            ObjectMetadata objectMetadata = new ObjectMetadata();
            objectMetadata.setContentLength(file.getSize());
            objectMetadata.setContentType(file.getContentType());

            try (InputStream inputStream = file.getInputStream()) {
                amazonS3.putObject(new PutObjectRequest(bucket, fileName, inputStream, objectMetadata)
                        .withCannedAcl(CannedAccessControlList.PublicRead));
            } catch (IOException e) {
                throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "파일 업로드에 실패했습니다.");
            }
            imageUrls.add(getUrl(bucket, fileName));
        });
        return imageUrls;
    }

    private GeneratePresignedUrlRequest createGeneratePresignedUrlRequest(String bucket, String filePath) {
        GeneratePresignedUrlRequest generatePresignedUrlRequest = new GeneratePresignedUrlRequest(bucket, filePath)
                .withMethod(HttpMethod.PUT)
                .withExpiration(getPreSignedUrlExpiration());
        generatePresignedUrlRequest.addRequestParameter(
                Headers.S3_CANNED_ACL,
                CannedAccessControlList.PublicRead.toString());
        return generatePresignedUrlRequest;
    }

    private Date getPreSignedUrlExpiration() {
        Date expiration = new Date();
        long expTimeMillis = expiration.getTime();
        expTimeMillis += 1000 * 60;
        expiration.setTime(expTimeMillis);
        return expiration;
    }

    private String createFileUuid() {
        return UUID.randomUUID().toString();
    }

    private String createFilePath(String prefix, String fileName) {
        String fileUuid = createFileUuid();
        return String.format("%s/%s-%s", prefix, fileUuid, fileName);
    }

    private String getUrl(String bucket, String fileName) {
        return amazonS3.getUrl(bucket, fileName).toString();
    }

}

File: src/main/java/com/zerozero/core/util/AWSS3Service.kt
package com.zerozero.core.util

import com.amazonaws.HttpMethod
import com.amazonaws.SdkClientException
import com.amazonaws.services.s3.AmazonS3
import com.amazonaws.services.s3.Headers
import com.amazonaws.services.s3.model.CannedAccessControlList
import com.amazonaws.services.s3.model.GeneratePresignedUrlRequest
import com.amazonaws.services.s3.model.ObjectMetadata
import com.amazonaws.services.s3.model.PutObjectRequest
import com.zerozero.image.exception.ImageErrorType
import com.zerozero.image.exception.ImageException
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.multipart.MultipartFile
import org.springframework.web.server.ResponseStatusException
import java.io.IOException
import java.net.MalformedURLException
import java.net.URL
import java.util.*
import java.util.function.Consumer

@Service
class AWSS3Service(
    @Value("\${cloud.aws.s3.bucket}")
    private val bucket: String,
    private val amazonS3: AmazonS3
) {

    fun getPreSignedUrl(prefix: String, fileName: String): String {
        val filePath = createFilePath(prefix, fileName)
        val generatePreSignedUrlRequest = createGeneratePreSignedUrlRequest(bucket, filePath)
        try {
            val url = amazonS3.generatePresignedUrl(generatePreSignedUrlRequest)
            return url.toString()
        } catch (e: SdkClientException) {
            throw ImageException(ImageErrorType.FAILED_TO_MAKE_URL)
        }
    }

    fun getObjectUrlFromPreSignedUrl(preSignedUrl: String): String {
        try {
            val url = URL(preSignedUrl)
            return "${url.protocol}://${url.host}${url.path}"
        } catch (e: MalformedURLException) {
            throw ImageException(ImageErrorType.FAILED_TO_MAKE_URL)
        }
    }

    fun uploadImage(prefix: String, multipartFile: MultipartFile): String {
        val fileName = createFilePath(prefix, multipartFile.originalFilename)

        val objectMetadata = ObjectMetadata()
        objectMetadata.contentLength = multipartFile.size
        objectMetadata.contentType = multipartFile.contentType

        amazonS3.putObject(bucket, fileName, multipartFile.inputStream, objectMetadata)
        return getUrl(bucket, fileName)
    }

    fun uploadImages(prefix: String, multipartFiles: List<MultipartFile>): List<String> {
        val imageUrls: MutableList<String> = ArrayList()

        multipartFiles.forEach(Consumer { file ->
            val fileName = createFilePath(prefix, file.originalFilename)
            val objectMetadata = ObjectMetadata()
            objectMetadata.contentLength = file.size
            objectMetadata.contentType = file.contentType

            try {
                file.inputStream.use { inputStream ->
                    amazonS3.putObject(
                        PutObjectRequest(bucket, fileName, inputStream, objectMetadata)
                            .withCannedAcl(CannedAccessControlList.PublicRead)
                    )
                }
            } catch (e: IOException) {
                throw ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "파일 업로드에 실패했습니다.")
            }
            imageUrls.add(getUrl(bucket, fileName))
        })
        return imageUrls
    }

    private fun createGeneratePreSignedUrlRequest(bucket: String?, filePath: String): GeneratePresignedUrlRequest {
        val generatePreSignedUrlRequest = GeneratePresignedUrlRequest(bucket, filePath)
            .withMethod(HttpMethod.PUT)
            .withExpiration(preSignedUrlExpiration())
        generatePreSignedUrlRequest.addRequestParameter(
            Headers.S3_CANNED_ACL,
            CannedAccessControlList.PublicRead.toString()
        )
        return generatePreSignedUrlRequest
    }

    private fun preSignedUrlExpiration(): Date {
        val expiration = Date()
        var expTimeMillis = expiration.time
        expTimeMillis += (1000 * 60).toLong()
        expiration.time = expTimeMillis
        return expiration
    }

    private fun createFileUuid(): String {
        return UUID.randomUUID().toString()
    }

    private fun createFilePath(prefix: String, fileName: String?): String {
        val safeFileName = fileName?.takeIf { it.isNotBlank() } ?: "file"
        val fileUuid = createFileUuid()
        return "$prefix/$fileUuid-$safeFileName"
    }


    private fun getUrl(bucket: String?, fileName: String): String {
        return amazonS3.getUrl(bucket, fileName).toString()
    }
}
